File: internal/parser/astjson/astjson.go
// Пакет astjson — спільний контракт «спрощений AST у JSON → ir». Будь-який
// фронтенд парсера (Python через ast, C++ через tree-sitter, браузер через
// Pyodide/web-tree-sitter) видає цей самий формат, а тут єдина конвертація в ir.
// Завдяки цьому конвертацію переиспользуют усі мови й працює вона і в WASM
// (без жодних підпроцесів).
package astjson

import (
	"encoding/json"
	"fmt"

	"flowgen/internal/ir"
)

// Node — вузол спрощеного дерева. Поля використовуються залежно від Kind:
//
//	process/io/call — Text
//	if              — Cond, Then, Else
//	for/while/dowhile — Cond, Body
//	block           — Stmts (вкладені block інлайняться)
type Node struct {
	Kind  string `json:"kind"`
	Text  string `json:"text"`
	Cond  string `json:"cond"`
	Stmts []Node `json:"stmts"`
	Then  *Node  `json:"then"`
	Else  *Node  `json:"else"`
	Body  *Node  `json:"body"`
}

// Func — іменована програма (функція або тіло модуля) → окрема схема.
type Func struct {
	Name  string `json:"name"`
	Block Node   `json:"block"`
}

// FromJSON розбирає JSON-масив Func і зводить його в список ir.Func. Це й є
// точка входу для WASM: на вхід — рядок від парсера, на вихід — готовий ir.
func FromJSON(data []byte) ([]ir.Func, error) {
	var fns []Func
	if err := json.Unmarshal(data, &fns); err != nil {
		return nil, fmt.Errorf("розбір AST-JSON: %w", err)
	}
	res := make([]ir.Func, 0, len(fns))
	for i := range fns {
		res = append(res, ir.Func{Name: fns[i].Name, Body: ToBlock(&fns[i].Block)})
	}
	return res, nil
}

// ToBlock зводить вузол-блок у ir.Block (вкладені block-вузли інлайняться).
func ToBlock(n *Node) *ir.Block {
	b := &ir.Block{}
	if n == nil {
		return b
	}
	for i := range n.Stmts {
		c := &n.Stmts[i]
		if c.Kind == "block" {
			b.Stmts = append(b.Stmts, ToBlock(c).Stmts...)
			continue
		}
		if node := ToNode(c); node != nil {
			b.Stmts = append(b.Stmts, node)
		}
	}
	return b
}

// ToNode зводить один вузол у ir.Node (nil — якщо тип невідомий).
func ToNode(n *Node) ir.Node {
	switch n.Kind {
	case "process":
		return &ir.Process{Text: n.Text}
	case "io":
		return &ir.IO{Text: n.Text}
	case "terminal":
		return &ir.Terminal{Text: n.Text}
	case "call":
		return &ir.Call{Text: n.Text}
	case "if":
		return &ir.If{Cond: n.Cond, Then: ToBlock(n.Then), Else: ToBlock(n.Else)}
	case "for":
		return &ir.For{Spec: n.Cond, Body: ToBlock(n.Body)}
	case "while":
		return &ir.While{Cond: n.Cond, Body: ToBlock(n.Body)}
	case "dowhile":
		return &ir.DoWhile{Cond: n.Cond, Body: ToBlock(n.Body)}
	}
	return nil
}

File: packages/engine/src/astjson.ts
// astjson — контракт «спрощений AST у JSON → IR». Порт pkg/parser/astjson.
// Будь-який фронтенд парсера (tree-sitter тощо) видає цей формат; тут єдина
// конвертація в IR. Вкладені block-вузли інлайняться.
import type { Block, Node, Func } from './ir.ts';

// AstNode — вузол спрощеного дерева. Поля залежать від kind (див. ToNode).
export interface AstNode {
  kind: string;
  text?: string;
  cond?: string;
  stmts?: AstNode[];
  then?: AstNode | null;
  else?: AstNode | null;
  body?: AstNode | null;
}
export interface AstFunc { name: string; block: AstNode; }

// fromJson: масив Func (рядок або вже розпарсений) → список IR-Func.
export function fromJson(data: string | AstFunc[]): Func[] {
  const fns: AstFunc[] = typeof data === 'string' ? JSON.parse(data) : data;
  return fns.map((f) => ({ name: f.name, body: toBlock(f.block) }));
}

// toBlock зводить вузол-блок у IR-Block (вкладені block інлайняться).
export function toBlock(n: AstNode | null | undefined): Block {
  const stmts: Node[] = [];
  for (const c of n?.stmts ?? []) {
    if (c.kind === 'block') stmts.push(...toBlock(c).stmts);
    else {
      const node = toNode(c);
      if (node) stmts.push(node);
    }
  }
  return { kind: 'block', stmts };
}

// toNode зводить один вузол у IR-Node (null — якщо тип невідомий).
function toNode(n: AstNode): Node | null {
  switch (n.kind) {
    case 'process': return { kind: 'process', text: n.text ?? '' };
    case 'io': return { kind: 'io', text: n.text ?? '' };
    case 'terminal': return { kind: 'terminal', text: n.text ?? '' };
    case 'call': return { kind: 'call', text: n.text ?? '' };
    case 'if': return { kind: 'if', cond: n.cond ?? '', then: toBlock(n.then), else: toBlock(n.else) };
    case 'for': return { kind: 'for', spec: n.cond ?? '', body: toBlock(n.body), else: toBlock(n.else) };
    case 'while': return { kind: 'while', cond: n.cond ?? '', body: toBlock(n.body), else: toBlock(n.else) };
    case 'dowhile': return { kind: 'dowhile', cond: n.cond ?? '', body: toBlock(n.body) };
    case 'infloop': return { kind: 'infloop', body: toBlock(n.body) };
    case 'break': return { kind: 'break' };
    case 'continue': return { kind: 'continue' };
  }
  return null;
}
